pic: initial support

afh opened the pull request:
This PR adds initial support for rendering a QR-Code as PIC, so that the QR-Code can be included in documents typeset using roff (only tested with GNU roff 1.23.0 so far).

The newly added code in `src/renders/pic.rs` is based on `src/renders/svg.rs` and was adapted as needed.
The example for pic includes additional files to show a full example including roff source. Within those files example commands are given to show a complete pipeline.

I'm not too familiar with Rust and hence kindly request a scrutinizing review. 😅

afh commented:
ℹ️ The generated output is compatible to what has been proposed to libqrencode in [`fukuchi/libqrencode#208`](https://github.com/fukuchi/libqrencode/pull/208)

kennytm commented on README.md:
```diff
@@ -144,3 +144,25 @@ Generates this output:
+
+Generates PIC output that renders as follows:
+
+[![Output](examples/encode_pic.png)](examples/encode_pic.png)
```
You should link to the text output of the PIC code, instead of the rendered PNG.

afh replied:
It would be possible to have all the call to the `p(…)` macro on a single line separated by `;`, would you prefer that?

kennytm replied:
> It would be possible to have all the call to the `p(…)` macro on a single line separated by `;`, would you prefer that?

if it is all in one line, there is no improvement in terms of readability or file size, so no.

kennytm commented on README.md:
````diff
@@ -144,3 +144,25 @@ Generates this output:
+}
+```
+
+Generates PIC output that renders as follows:
````
PIC seems like some obscure markup format unlike PNG or SVG. I suppose it is https://en.wikipedia.org/wiki/PIC_(markup_language)? Please provide a link to introduce what PIC is.

afh replied:
Correct that's the PIC this PR adds support for.
Would linking to a PDF be acceptable?
I found ["PIC — A Graphics Language for Typesetting"](https://archive.org/details/pic-graphics-language/mode/2up) the best source material.

kennytm replied:
The primary introduction should not be this PDF file, in the same way if I were to introduce SVG I wouldn't link to https://www.w3.org/TR/SVG11/. 

This PDF file is a nice addition as a "for more details" link.

kennytm commented on src/render/pic.rs:
```diff
@@ -0,0 +1,78 @@
+        Canvas {
+            pic: format!(
+                concat!(
+                    r#"command ".if !mqrcode_fg .defcolor qrcode_fg rgb {fg}""#,
+                    "\n",
+                    r#"command ".if !mqrcode_bg .defcolor qrcode_bg rgb {bg}""#,
+                    "\n",
```
I don't see any point in supporting overriding the color externally. Remove it.

```suggestion
                    "qrcode_fg = {fg}\n",
                    "qrcode_bg = {bg}\n",
```

afh replied:
ℹ️ GNU PIC does not support defining colors on its own, but rather relies on the color definition capabilities of GNU roff.

To remove support for overriding colors only the `.if !mqrcode_[fb]g` needs to be removed.

I'll address this shortly.

afh replied:
The point of being able to override the color is to be able to use the same QR-Code source PIC file and change the color to match the theme of the document it is included in without the need to re-generate the QR-Code or store other generated  version of the same QR-Code in separate files.

What are reasons against supporting it?

kennytm replied:
The content of the `command` command is post-processor-dependent. Using `.if` makes the generated file impossible to be used in TeX mode or standalone implementations like pikchr. 

I don't find "change the color to match the theme of the document" to be a good argument to special-case PIC here. The same argument applies to SVG and PNG after all. If you really need these special treatment you could always prepend these magic commands before writing the file, or edit the file after it is generated.

afh replied:
Excellent objection, @kennytm, thank you very much for that 😃
I had not considered the portability of the generated PIC code!

At first glance it appears to me that [pikchr](https://pikchr.org) takes much inspiration from PIC, but makes not guarantees on compatibility.

Since "Colored Objects [are] a GNU extension" (see section "6.10 Colored Objects" in ["Making Pictures with GNU PIC"](https://archive.org/details/manualzilla-id-5804556)) and pikchr has a slightly different syntax for colored objects I suggest to remove color support altogether.

What do you think, @kennytm?

kennytm replied:
> Since "Colored Objects [are] a GNU extension" (see section "6.10 Colored Objects" in ["Making Pictures with GNU PIC"](https://archive.org/details/manualzilla-id-5804556)) and pikchr has a slightly different syntax for colored objects I suggest to remove color support altogether.

that's fine for me :)

kennytm commented on src/render/pic.rs:
```diff
@@ -0,0 +1,78 @@
+                    r#"command ".if !mqrcode_bg .defcolor qrcode_bg rgb {bg}""#,
+                    "\n",
+                    "maxpswid={w};maxpsht={h};movewid=0;moveht=1;boxwid=1;boxht=1\n",
+                    r#"define P {{ box wid $3 ht $4 color "qrcode_fg" with .nw at $1,-$2 }}"#,
```
```suggestion
                    r#"define p {{ box wid $3 ht $4 fill qrcode_fg with .nw at $1,-$2 }}"#,
```

"Variable names begin with a lower case letter"

kennytm commented on src/render/pic.rs:
```diff
@@ -0,0 +1,78 @@
+                    "maxpswid={w};maxpsht={h};movewid=0;moveht=1;boxwid=1;boxht=1\n",
+                    r#"define P {{ box wid $3 ht $4 color "qrcode_fg" with .nw at $1,-$2 }}"#,
+                    "\n",
+                    r#"box wid maxpswid ht maxpsht color "qrcode_bg" with .nw at 0,0"#,
```
```suggestion
                    r#"box wid maxpswid ht maxpsht fill qrcode_bg with .nw at 0,0"#,
```

afh replied:
The color names must be quoted, because AFAIK PIC does not offer arbitrary definition of variables.

kennytm replied:
okay

actually I find that the color specification is conflicting among different dialects, could you make it

```pic
box wid maxpswid ht maxpsht {bg} with .nw at 0,0
```

and make the usage `pic::Color("color \"#ff0000\"")` or `pic::Color("fill 0xff0000")` 🤔

afh replied:
While I was writing [this comment](https://github.com/kennytm/qrcode-rust/pull/65#discussion_r1661818111) you wrote the one above 🙂 

It seems you also prefer to remove support for colors. I'll go ahead and make the necessary changes.

kennytm commented on src/render/pic.rs:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    fn draw_dark_rect(&mut self, left: u32, top: u32, width: u32, height: u32) {
+        write!(self.pic, "P({left},{top},{width},{height})\n").unwrap();
```
```suggestion
        write!(self.pic, "p({left},{top},{width},{height})\n").unwrap();
```

afh commented:
Kindly requesting re-review, @kennytm. The current state of this PR and the PIC code it generates offers to most compatible output according to my tests. Hopefully I've addressed all of your other comments properly, if not please let me know 🙂

kennytm commented on README.md:
````diff
@@ -144,3 +144,247 @@ Generates this output:
+Generates [PIC](https://en.wikipedia.org/wiki/PIC_(markup_language))
+output that renders as follows:
+
+```pic
+maxpswid=29;maxpsht=29;movewid=0;moveht=1;boxwid=1;boxht=1
+define p { box wid $3 ht $4 fill 1 with .nw at $1,-$2 }
+box wid maxpswid ht maxpsht with .nw at 0,0
+p(4,4,1,1)
+p(5,4,1,1)
+p(6,4,1,1)
+p(7,4,1,1)
````
well let's not include 200 lines of code here either 😂 

link to the `test_annex_i_qr_as_pic.pic` file, or truncate the file end it with `…`.

afh replied:
I assume include one _very_ long line of code isn't very helpful either 😉 hence I went ahead and truncated the code (including `…` at the end` and linked to `test_annex_i_qr_as_pic.pic` below the truncated example.

kennytm commented on src/render/pic.rs:
```diff
@@ -0,0 +1,72 @@
+
+/// A PIC color.
+#[derive(Copy, Clone, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Color<'a>(pub &'a str);
```
is the `&str` member still needed?

afh replied:
I'm afraid answering that question exceeds my knowledge of Rust 🙁 I'd say "No, it's no longer needed", but trying to remove it results in compiler errors I'm unable to resolve. Would you be willing to help out here, @kennytm?

kennytm replied:
you could make it a unit struct

```rust
pub struct Color;
```

and remove all mentions of `'a` in this file.

afh replied:
Thanks for the helpful suggestion, @kennytm, I've made the appropriate changes to this PR.

kennytm commented on src/render/pic.rs:
```diff
@@ -0,0 +1,72 @@
+#[doc(hidden)]
+pub struct Canvas {
+    pic: String,
+    marker: PhantomData<Color>,
```
```suggestion
```

no need to include the PhantomData since it no longer needs to capture the unused generic argument.

kennytm reviewed (commented):
Rest LGTM

kennytm commented:
and please fix the `rustfmt` before committing, running `cargo fmt` should be enough.

afh commented:
Thank you for your patience and helpful advice on this, @kennytm, very much appreciated! 🙏

I've addressed your comments in the latest force-push to this PR.

kennytm commented on src/render/pic.rs:
```diff
@@ -0,0 +1,69 @@
+            pic: format!(
+                concat!(
+                    "maxpswid={w};maxpsht={h};movewid=0;moveht=1;boxwid=1;boxht=1\n",
+                    r#"define p {{ box wid $3 ht $4 fill 1 with .nw at $1,-$2 }}"#,
+                    "\n",
+                    r#"box wid maxpswid ht maxpsht with .nw at 0,0"#,
+                    "\n",
```
```suggestion
                    "define p {{ box wid $3 ht $4 fill 1 with .nw at $1,-$2 }}\n",
                    "box wid maxpswid ht maxpsht with .nw at 0,0\n",
```

afh commented:
Thanks for taking another scrutinising look, suggesting helpful changes, and already approving this PR. I've updated this PR to include your suggestions 🙂

afh commented:
Thanks for your patience, guidance and help on getting this merged, kennytm, _very_ much appreciated! 😃 For this new functionality to be useful for me a new release would be helpful, so that the `qrcode-rust` dependency in [sorairolake/qrtool](https://github.com/sorairolake/qrtool) can be updated and a PR to add the necessary changes to support PIC output can created.

Are there any plans for a new release any time soon(ish) by any chance?
